avl_tree/Cpp/avl_tree.cpp

```diff
@@ -0,0 +1,392 @@
+int heighe	  - Height of the node
+*/
+class node {
+	public:
+	/*
+	Constructor for node
+	Parameters:
+	int v - sets value of node to v.
+	 */
+
+	explicit node(int v) {
+		this->val = v;
+		this->left = NULL;
+		this->right = NULL;
+		this->height = 1;
+	}
+
+	/*
+	Getter functions
+	*/
+	int get_val() {
+		return val;
+	}
+
+	node * get_left() {
+		return left;
+	}
+
+	node * get_right() {
+		return right;
+	}
+
+	int get_height() {
+		return height;
+	}
+
+	/*
+	Function to return balance factor of a node
+
+	Parameters:
+	node * t  -  Pointer to node
+
+	Returns:
+	int : balance factor of node t.
+	*/
+	int get_bal_fact() {
+		int ht_lt = get_left() ? get_left()->get_height() : 0;
+		int ht_rt = get_right() ? get_right()->get_height() : 0;
+		return (ht_lt-ht_rt);
+	}
+
+	node * right_rotate() {
+		if(this->get_left() == NULL) return this;
+		node * lt = this->get_left();
+		node * tmp = lt->get_right();
+		lt->set_right(this);
+		this->set_left(tmp);
+
+		int ht_lt = this->get_left() ? this->get_left()->get_height() : 0;
+		int ht_rt = this->get_right() ? this->get_right()->get_height() : 0;
+
+		this->set_height(max(ht_lt, ht_rt) + 1);
+
+		int ht_lt_1 = lt->get_left() ? lt->get_left()->get_height() : 0;
+		int ht_rt_1 = lt->get_right() ? lt->get_right()->get_height() : 0;
+
+		lt->set_height(max(ht_lt_1, ht_rt_1) + 1);
+
+		return lt;
+	}
+
+	node * left_rotate() {
+		if(this->get_right() == NULL) return this;
+		node * rt = this->get_right();
+		node * tmp = rt->get_left();
+
+		rt->set_left(this);
+		this->set_right(tmp);
+
+
+		int ht_lt = this->get_left() ? this->get_left()->get_height() : 0;
+		int ht_rt = this->get_right() ? this->get_right()->get_height() : 0;
+
+		this->set_height(max(ht_lt, ht_rt) + 1);
+
+		int ht_lt_1 = rt->get_left() ? rt->get_left()->get_height() : 0;
+		int ht_rt_1 = rt->get_right() ? rt->get_right()->get_height() : 0;
+
+		rt->set_height(max(ht_lt, ht_rt) + 1);
+
+		return rt;
+	}
+
+
+	/*
+	Setter functions
+	*/
+	void set_val(int v) {
+		this->val = v;
+	}
+
+	void set_left(node * l) {
+		this->left = l;
+	}
+
+	void set_right(node * r) {
+		this->right = r;
+	}
+
+	void set_height(int h) {
+		this->height = h;
+	}
+
+	private:
+	int val;
+	node * left;
+	node * right;
+	int height;
+};
+
+
+/*
+Class for defining AVL tree
+
+Has -
+node * root - Node pointer to the root of the tree
+*/
+class AVL {
+	public:
+	/*
+	Class' public member functions
+	*/
+	AVL() {
+		this->root = NULL;
+	}
+
+	node * search(int v) {
+		return search_util(root, v);
+	}
+
+	void insert(int v) {
+		root = insert_util(root, v);
+	}
+
+	void inorder_traversal() {
+		inorder_util(root);
+	}
+
+	void preorder_traversal() {
+		preorder_util(root);
+	}
+
+	void postorder_traversal() {
+		postorder_util(root);
+	}
+
+	void delete_node(int v) {
+		root = delete_node_util(root, v);
+	}
+
+	private:
+	node * root;
+
+	/*
+	Class' private member functions
+	*/
+	node * search_util(node * t, int v) {
```

Comment:
what is `t` and `v` is missing in the doc string

Reply:
Done 👍